File: data/EXP01/python/2043/2043.output_3.py
# @lc app=leetcode id=2043 lang=python3
class Bank:

    def __init__(self, balance: List[int]):
        self.balance = balance

    def transfer(self, account1: int, account2: int, money: int) -> bool:
        if self._is_valid_account(account1) and self._is_valid_account(account2) and money <= self.balance[account1-1]:
            self.balance[account1-1] -= money
            self.balance[account2-1] += money
            return True
        return False

    def deposit(self, account: int, money: int) -> bool:
        if self._is_valid_account(account):
            self.balance[account-1] += money
            return True
        return False

    def withdraw(self, account: int, money: int) -> bool:
        if self._is_valid_account(account) and money <= self.balance[account-1]:
            self.balance[account-1] -= money
            return True
        return False

    def _is_valid_account(self, account: int) -> bool:
        return account >= 1 and account <= len(self.balance)

File: data/EXP01/golang_no_system_var/2043/2043.output_3.go
type Bank struct {
    balance []int64
}

func Constructor(balance []int64) Bank {
    return Bank{
        balance: balance,
    }
}

func (b *Bank) Transfer(account1 int, account2 int, money int64) bool {
    if !b.validateAccount(account1) || !b.validateAccount(account2) {
        return false
    }

    if b.balance[account1-1] < money {
        return false
    }

    b.balance[account1-1] -= money
    b.balance[account2-1] += money

    return true
}

func (b *Bank) Deposit(account int, money int64) bool {
    if !b.validateAccount(account) {
        return false
    }

    b.balance[account-1] += money
    return true
}

func (b *Bank) Withdraw(account int, money int64) bool {
    if !b.validateAccount(account) {
        return false
    }

    if b.balance[account-1] < money {
        return false
    }

    b.balance[account-1] -= money
    return true
}

func (b *Bank) validateAccount(account int) bool {
    if account < 1 || account > len(b.balance) {
        return false
    }
    return true
}
